go/vt/vttablet/tabletserver/throttle/base/self_metric_loadavg.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2024 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package base
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"sync/atomic"
+
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/connpool"
+)
+
+var _ SelfMetric = registerSelfMetric(&LoadAvgSelfMetric{})
+
+type LoadAvgSelfMetric struct {
+	hostCpuCoreCount atomic.Int32
+}
+
+func (m *LoadAvgSelfMetric) Name() MetricName {
+	return LoadAvgMetricName
+}
+
+func (m *LoadAvgSelfMetric) DefaultScope() Scope {
+	return SelfScope
+}
+
+func (m *LoadAvgSelfMetric) DefaultThreshold() float64 {
+	return 1.0
+}
+
+func (m *LoadAvgSelfMetric) RequiresConn() bool {
+	return false
+}
+
+func (m *LoadAvgSelfMetric) Read(ctx context.Context, throttler ThrottlerMetricsPublisher, conn *connpool.Conn) *ThrottleMetric {
+	metric := &ThrottleMetric{
+		Scope: SelfScope,
+	}
+
+	coreCount := m.hostCpuCoreCount.Load()
+	if coreCount == 0 {
+		// Count cores. This number is not going to change in the lifetime of this tablet,
+		// hence it makes sense to read it once then cache it.
+
+		// We choose to read /proc/cpuinfo over executing "nproc" or similar commands.
+		var coreCount int32
+		f, err := os.Open("/proc/cpuinfo")
```

Comment:
This means that throttling based on load avg is only supported on linux, right? It might be good to check for file not found errors and return a message about this (or even just check the os name first).

Even better would be to leverage the go runtime package instead to eliminate the platform requirement: `runtime.NumCPU()`

The other advantage that has is it returns the number of hardware threads that are usable by the process. The operator may have limited the number of threads that the process can use via `taskset` etc.

Reply:
Fixed. Now using `runtime.NumCPU()` (Doh! How did I not use it beforehand??) as well as `runtime.GOOS` and returning error metric when OS is not `"linux"`.

Reply:
Now that we're not using `/proc/cpuinfo`, I don't think we need an OS check — it should work fine on macOS, *BSD, even Windows. Unless there's some other aspect of the general load avg code that assumes/requres linux.

Reply:
oh, we still read from `/proc/loadavg`